src/frequenz/sdk/timeseries/formulas/_ast.py

```diff
@@ -0,0 +1,205 @@
+# License: MIT
+# Copyright © 2025 Frequenz Energy-as-a-Service GmbH
+
+"""Formula AST nodes and evaluation logic."""
+
+from __future__ import annotations
+
+import abc
+import logging
+import math
+from collections.abc import AsyncIterator
+from dataclasses import dataclass
+from typing import Generic
+
+from typing_extensions import override
+
+from ..._internal._math import is_close_to_zero
+from .._base_types import QuantityT, Sample
+from ._functions import Function
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Node(abc.ABC):
+    """An abstract syntax tree node representing a formula expression."""
+
+    span: tuple[int, int] | None
+
+    @abc.abstractmethod
+    def evaluate(self) -> float | None:
+        """Evaluate the expression and return its numerical value."""
+
+    @abc.abstractmethod
+    def format(self, wrap: bool = False) -> str:
+        """Return a string representation of the node."""
+
+    @override
+    def __str__(self) -> str:
+        """Return the string representation of the node."""
+        return self.format()
+
+
+@dataclass
+class TelemetryStream(Node, Generic[QuantityT]):
+    """A AST node that retrieves values from a component's telemetry stream."""
+
+    source: str
```

Comment:
The properties are missing docs, I think mkdocs will not show them if they are undocumented (also it is not clear to me what `source` is).